### PR TITLE
feat(Input): make `label` prop optional & add `accessibilityLabel` prop

### PR DESCRIPTION
**Pull request by snitin315**
Fixes https://github.com/razorpay/blade/issues/1237

I'm thrilled to introduce an important update to our TextInput component, aimed at improving accessibility and flexibility. With this latest release, we have made the `label` prop optional while introducing the all-new `accessibilityLabel` prop, ensuring an inclusive user experience for everyone.

### Key Updates:

**Optional `label` Prop**: We understand that not all use cases require a label for the TextInput. Therefore, we have made the label prop optional, providing you with the freedom to choose whether to display a label or not, depending on your specific application requirements.

**Introducing `accessibilityLabel`:** Recognizing the significance of accessibility in modern applications, we have added the `accessibilityLabel` prop to the TextInput. This prop enables developers to assign a descriptive label for the input field, making it more user-friendly for individuals using assistive technologies or screen readers.

**Enhanced User Guidance:** To maintain usability, we have implemented a requirement that either the `label` or `accessibilityLabel` prop must be provided. This ensures that users will always have clear guidance when interacting with TextInput, promoting a seamless user experience.

**Updated Stories:** As part of this PR, I have also added new stories for the TextInput without a label, demonstrating its versatility and how it can be used effectively without a visible label.

### CODE REVIEW

File: packages/blade/src/components/Input/TextInput/TextInput.tsx
Comment (anuraghazra) [inline packages/blade/src/components/Input/TextInput/TextInput.tsx:231]: Lets reduce runtime checks as much as possible, most of the time we do runtime checks for things which can't be easily done with TypeScript. 

But this label vs accessibilityLabel case can be done via TypeScript, we are doing the exact same thing in BaseButton also with children vs icon. 


```ts

type TextInputCommonProps = Pick<
  BaseInputProps,
  ...
>

/*
  Mandatory accessibilityLabel prop when label is not provided
*/
type TextInputPropsWithLabel = {
  /**
   * Label to be shown for the input field
   */
  label?: undefined;
  /**
   * Accessibility label for the input
   */
  accessibilityLabel: string;
};

/*
  Optional accessibilityLabel prop when label is provided
*/
type TextInputPropsWithA11yLabel = {
  /**
   * Label to be shown for the input field
   */
  label: string;
  /**
   * Accessibility label for the input
   */
  accessibilityLabel?: string;
};

type TextInputProps = (TextInputPropsWithA11yLabel | TextInputPropsWithLabel) &
  TextInputCommonProps;
```

image

image

image

> if you are planning to support this on all the components you can move these types to the BaseInput instead

Comment (snitin315) [reply packages/blade/src/components/Input/TextInput/TextInput.tsx:231]: @anuraghazra but if someone is using Blade with JS they will bypass this check. or by adding `//@ts-ignore` with TS.

Comment (snitin315) [reply packages/blade/src/components/Input/TextInput/TextInput.tsx]: Updated.

Comment (anuraghazra) [reply packages/blade/src/components/Input/TextInput/TextInput.tsx]: > but if someone is using Blade with JS they will bypass this check. or by adding //@ts-ignore with TS.

That's fine. let them bypass. 

```
<labe>
  My custom label
  <Input accessiblityLabel="My custom label" /> // this is redundant can i please ignore this? the label is already set
</label>
```

File: .changeset/nervous-chicken-explain.md
Comment (snitin315) [inline .changeset/nervous-chicken-explain.md:13]:
```suggestion
- **Introducing `accessibilityLabel`:** Recognizing the significance of accessibility in modern applications, we have added the `accessibilityLabel` prop to the Input components. This prop enables developers to assign a descriptive label for the input field, making it more user-friendly for individuals using assistive technologies or screen readers.

- **Enhanced User Guidance:** To maintain usability, we have implemented a requirement that either the `label` or `accessibilityLabel` prop must be provided. This ensures that users will always have clear guidance when interacting with Inputs, promoting a seamless user experience.
```

File: .changeset/nervous-chicken-explain.md
Comment (snitin315) [inline .changeset/nervous-chicken-explain.md:11]:
```suggestion
- **Introducing `accessibilityLabel`:** Recognizing the significance of accessibility in modern applications, we have added the `accessibilityLabel` prop to the `TextInput`, `TextArea`, `PasswordInput`, `SelectInput`, and `OTPInput` components. This prop enables developers to assign a descriptive label for the input field, making it more user-friendly for individuals using assistive technologies or screen readers.
```

File: .changeset/nervous-chicken-explain.md
Comment (snitin315) [inline .changeset/nervous-chicken-explain.md:5]:
```suggestion
feat(Input): make `label` prop optional & add `accessibilityLabel` prop to `TextInput`, `TextArea`, `PasswordInput`, `SelectInput`, and `OTPInput` components
```

File: .changeset/nervous-chicken-explain.md
Comment (snitin315) [inline .changeset/nervous-chicken-explain.md:9]:
```suggestion
- **Optional `label` Prop**: We understand that not all use cases require a label for the Input components. Therefore, we have made the label prop optional, providing you with the freedom to choose whether to display a label or not, depending on your specific application requirements.
```

**Review by anuraghazra — DISMISSED** [review]
LGTM, assuming you've tested the TS error on all the inputs

**Comment by snitin315** [comment]
@anuraghazra It seems updating only BaseInput types was not sufficient. Updated types for all inputs in https://github.com/razorpay/blade/pull/1472/commits/62786d7ff24da58141fad0a59e366308bb374db8.

Screenshot 2023-07-27 at 8 22 46 AM

**Comment by anuraghazra** [comment]
> It seems updating only BaseInput types was not sufficient.

Hmm wierd, ideally should work I'll look into this.

File: packages/blade/src/components/Input/OTPInput/OTPInput.tsx
Comment (saurabhdaware) [inline packages/blade/src/components/Input/OTPInput/OTPInput.tsx:380]: why is `as string` needed?

File: packages/blade/src/components/Input/BaseInput/BaseInput.tsx
Comment (snitin315) [reply packages/blade/src/components/Input/BaseInput/BaseInput.tsx:276]: Yeah, I think the naming got the opposite.

File: packages/blade/src/components/Input/OTPInput/OTPInput.tsx
Comment (snitin315) [reply packages/blade/src/components/Input/OTPInput/OTPInput.tsx:380]: Screenshot 2023-07-27 at 4 54 10 PM

File: packages/blade/src/components/Input/BaseInput/BaseInput.tsx
Comment (snitin315) [reply packages/blade/src/components/Input/BaseInput/BaseInput.tsx]: Fixed in https://github.com/razorpay/blade/pull/1472/commits/2b1141af3dc71065752ceee411f9ace6fbe8396c

File: .changeset/nervous-chicken-explain.md
Comment (snitin315) [reply .changeset/nervous-chicken-explain.md:11]: 100% use of brain.

Comment (nashcheez) [reply .changeset/nervous-chicken-explain.md:11]: @snitin315 should be banned at using ChatGPT!

File: packages/blade/src/components/Input/OTPInput/OTPInput.tsx
Comment (saurabhdaware) [reply packages/blade/src/components/Input/OTPInput/OTPInput.tsx:380]: Can we make the FormLabel accept string or undefined then?

Comment (snitin315) [reply packages/blade/src/components/Input/OTPInput/OTPInput.tsx]: @saurabhdaware done in https://github.com/razorpay/blade/pull/1472/commits/89956db0570493fed65157842048ed6566fb0ad4

**Review by saurabhdaware — APPROVED** [review]
Looks good to me. Do accept the chromatic changes before merging.

**Pull request merged**
